pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,517 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	operatorcontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	nodeutils "github.com/aws/karpenter-core/pkg/utils/node"
+)
+
+// Controller is the deprovisioning controller.
+type Controller struct {
+	kubeClient             client.Client
+	cluster                *state.Cluster
+	provisioner            *provisioning.Provisioner
+	recorder               events.Recorder
+	clock                  clock.Clock
+	cloudProvider          cloudprovider.CloudProvider
+	emptiness              *Emptiness
+	consolidation          *Consolidation
+	lastConsolidationState int64
+}
+
+// pollingPeriod that we inspect cluster to look for opportunities to deprovision
+const pollingPeriod = 10 * time.Second
+
+var errCandidateNodeDeleting = fmt.Errorf("candidate node is deleting")
+
+// waitRetryOptions are the retry options used when waiting on a node to become ready or to be deleted
+// readiness can take some time as the node needs to come up, have any daemonset extended resoruce plugins register, etc.
+// deletion can take some time in the case of restrictive PDBs that throttle the rate at which the node is drained
+var waitRetryOptions = []retry.Option{
+	retry.Delay(2 * time.Second),
+	retry.LastErrorOnly(true),
+	retry.Attempts(60),
+	retry.MaxDelay(10 * time.Second), // 22 + (60-5)*10 =~ 9.5 minutes in total
+}
+
+func NewController(clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
+	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Controller {
+	return &Controller{
+		clock:         clk,
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		provisioner:   provisioner,
+		recorder:      recorder,
+		cloudProvider: cp,
+		emptiness:     &Emptiness{kubeClient: kubeClient, clock: clk, cluster: cluster},
+		consolidation: NewConsolidation(clk, kubeClient, provisioner, cp, recorder, cluster),
+	}
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) operatorcontroller.Builder {
+	return operatorcontroller.NewSingletonManagedBy(m).
+		Named("deprovisioning")
+}
+
+func (c *Controller) LivenessProbe(_ *http.Request) error {
+	return nil
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// the last cluster consolidation wasn't able to improve things and nothing has changed regarding
+	// the cluster that makes us think we would be successful now
+	if c.lastConsolidationState == c.cluster.ClusterConsolidationState() {
```

Comment:
If we have this check, is it worth waiting 10 seconds? Can we check this every 100ms?

Reply:
This was from a previous change. Not changing anything right now.